### PR TITLE
C5T1-27 Initial interface for compression

### PR DESCRIPTION
**Pull request by klinvill**
This PR adds an initial interface and data structures for the compression functionality (using flate2). It also adds comments to the main function describing the general program flow. 

One of the questions this work raises is if we should include compression at all. Compression support is removed in TLS 1.3 since compressing information before encrypting it can potentially leak information. I think it's ok to have compression for our project (along with a disclaimer that compressing info can lead to leaks) since it's just a class project. Since TLS 1.3 does not support compression, I was thinking we could just add our own custom compression layer to start. After that, if we wanted to, we could add compression support for TLS 1.2 or for particular applications (like HTTP) even over TLS 1.3.

### CODE REVIEW

**Review by made9505 — APPROVED** [review]

**Comment by made9505** [comment]
Will need to read more about rust syntax and features, but approve for now.

**Comment by Uyutaka** [comment]
I also need to read more, but I approve for now.

**Review by Uyutaka — APPROVED** [review]

File: src/main.rs [hunk 1]
```diff
@@ -1,64 +1,28 @@
...
+mod compression;
+
+fn main() {
+    // General flow:
+    //  - parse command line arguments:
+    //      - run as forward or reverse proxy depending on args
+    //      - optionally compress data before encrypting (vulnerable to CRIME-style attacks)
+    //  - accept new TCP connections, handle each one in a new thread
+    //
+    //  - If forward proxy (runs for each new connection):
```
Comment (Uyutaka) [inline src/main.rs:10]: Does "forward proxy" mean client-router in our network?
And, "reverse proxy" means server-router?

Comment (klinvill) [reply src/main.rs:10]: Yup, in our case that's right. A good explanation of the terms is here: https://www.cloudflare.com/learning/cdn/glossary/reverse-proxy/

Comment (Uyutaka) [reply src/main.rs:10]: Thanks

**Pull request merged**